packages/protocol/contracts/layer2/based/ShastaAnchor.sol

```diff
@@ -0,0 +1,343 @@
+    /// @param _provabilityBondGwei The provability bond amount in Gwei.
+    /// @param _signalService The address of the signal service.
+    /// @param _pacayaForkHeight The block height at which the Pacaya fork is activated.
+    /// @param _shastaForkHeight The block height at which the Shasta fork is activated.
+    /// @param _syncedBlockManager The address of the synced block manager.
+    /// @param _bondManager The address of the bond manager.
+    constructor(
+        uint48 _livenessBondGwei,
+        uint48 _provabilityBondGwei,
+        address _signalService,
+        uint64 _pacayaForkHeight,
+        uint64 _shastaForkHeight,
+        ISyncedBlockManager _syncedBlockManager,
+        IShastaBondManager _bondManager
+    )
+        PacayaAnchor(_signalService, _pacayaForkHeight, _shastaForkHeight)
+    {
+        require(
+            _shastaForkHeight == 0 || _shastaForkHeight > _pacayaForkHeight, InvalidForkHeight()
+        );
+
+        livenessBondGwei = _livenessBondGwei;
+        provabilityBondGwei = _provabilityBondGwei;
+        syncedBlockManager = _syncedBlockManager;
+        bondManager = _bondManager;
+    }
+
+    // ---------------------------------------------------------------
+    // External functions
+    // ---------------------------------------------------------------
+
+    /// @notice Processes a block within a proposal, handling bond instructions and L1 data
+    /// anchoring.
+    /// @dev Core function that processes blocks sequentially within a proposal:
+    ///      1. Designates prover on first block (blockIndex == 0)
+    ///      2. Processes bond transfers with cumulative hash verification
+    ///      3. Anchors L1 block data for cross-chain verification
+    ///      4. Tracks parent block hash to prevent duplicate calls
+    /// @param _proposalId Unique identifier of the proposal being anchored.
+    /// @param _proposer Address of the entity that proposed this batch of blocks.
+    /// @param _proverAuth Encoded ProverAuth for prover designation (empty after block 0).
+    /// @param _bondInstructionsHash Expected cumulative hash after processing instructions.
+    /// @param _bondInstructions Bond credit instructions to process for this block.
+    /// @param _blockIndex Current block index within the proposal (0-based).
+    /// @param _anchorBlockNumber L1 block number to anchor (0 to skip anchoring).
+    /// @param _anchorBlockHash L1 block hash at _anchorBlockNumber.
+    /// @param _anchorStateRoot L1 state root at _anchorBlockNumber.
+    /// @return isLowBondProposal_ True if proposer has insufficient bonds.
+    /// @return designatedProver_ Address of the designated prover.
+    function updateState(
+        // Proposal level fields - define the overall batch
+        uint48 _proposalId,
+        address _proposer,
+        bytes calldata _proverAuth,
+        bytes32 _bondInstructionsHash,
+        LibBonds.BondInstruction[] calldata _bondInstructions,
```

Comment:
How are the values of `_bondInstructions` derived? Is it from the `emit BondInstructed` event of the proposal being processed?

Reply:
One thing to consider compatibility with preconfs: If the anchor execution depends on finalizations of past blocks at proposal time, at preconf time, you don't know exactly which `_bindInstructions` have to be executed, which makes it hard to preconf the exact L2 state.